CAPABILITIES.md

```diff
@@ -9,59 +9,78 @@ and other identifiers that are supported by Conscrypt.
-but will ignore it.
+* `SSLv3` (ignored)
+* `TLSv1`
+* `TLSv1.1`
+* `TLSv1.2`
+* `TLSv1.3`
+
+Conscrypt supports TLS v1.0-1.3.  For backwards compatibility it will accept
+`SSLv3` in calls to methods like
+[`setEnabledProtocols()`](https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLSocket.html#setEnabledProtocols-java.lang.String:A-)
+but will ignore it.  TLS 1.3 is not enabled unless `SSLContext.TLSv1.3` is
+requested or it's explicitly enabled by a call to
+[`setEnabledProtocols()`](https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLSocket.html#setEnabledProtocols-java.lang.String:A-).
 
 ### SSLContext
 
-* Default
-* SSL
-* TLS
-* TLSv1
-* TLSv1.1
-* TLSv1.2
-
-Conscrypt provides the above set of SSLContext algorithm names for JSSE purposes, including
-the special value "Default", which is used to determine the value of
-[SSLContext.getDefault()](https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLContext.html#getDefault--).
-All of these values return a context where TLS v1.0-1.2 are all enabled.
+* `Default`
+* `SSL`
+* `TLS`
+* `TLSv1`
+* `TLSv1.1`
+* `TLSv1.2`
+* `TLSv1.3`
+
+Conscrypt provides the above set of SSLContext algorithm names for JSSE
+purposes, including the special value `Default`, which is used to determine the
+value of
+[`SSLContext.getDefault()`](https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLContext.html#getDefault--).
+All of these values return a context where TLS v1.0-1.2 are all enabled except
+`TLSv1.3`, which has TLS v1.0-1.3 enabled.
 
 ### Cipher Suites
 
 #### Enabled
-* TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
-* TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
-* TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
-* TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
-* TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
-* TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
-* TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
-* TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
-* TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
-* TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
-* TLS_RSA_WITH_AES_128_CBC_SHA
-* TLS_RSA_WITH_AES_128_GCM_SHA256
-* TLS_RSA_WITH_AES_256_CBC_SHA
-* TLS_RSA_WITH_AES_256_GCM_SHA384
-
-The above cipher suites are enabled by default.
+* TLS 1.0-1.2
+  * `TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA`
+  * `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256`
+  * `TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA`
+  * `TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384`
+  * `TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256`
+  * `TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA`
+  * `TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256`
+  * `TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA`
+  * `TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384`
+  * `TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256`
+  * `TLS_RSA_WITH_AES_128_CBC_SHA`
+  * `TLS_RSA_WITH_AES_128_GCM_SHA256`
+  * `TLS_RSA_WITH_AES_256_CBC_SHA`
+  * `TLS_RSA_WITH_AES_256_GCM_SHA384`
+* TLS 1.3
+  * `TLS_AES_128_GCM_SHA256`
+  * `TLS_AES_256_GCM_SHA384`
+  * `TLS_CHACHA20_POLY1305_SHA256`
+
+The above cipher suites are enabled by default when the associated version of
```

Comment:
Is this part of TLS 1.3 spec? or a backwards compatibility feature? Seems surprising given potential future need to disable an insecure cipher, or reorder for efficiency reasons on different platforms.

I removed TLS 1.3 from the default connection spec in OkHttp because ciphers wouldn't work anyway, but with this looks like they would.

Reply:
This is a design decision on our part, though it meshes well with the overall philosophy of TLS 1.3.  The idea is that by only including a small number of well-supported, highly-tested cipher suites, there should be no need to customize the cipher suites, and that makes life simpler for everyone.  It also allows us more leeway to adjust preference orderings based on things like whether the platform being run on has hardware-accelerated AES or drop old ciphers without causing problems for callers.

(As background, TLS 1.3 only defines five cipher suites total, of which we support three, the two AES/GCM ones and the ChaCha20 one.  A significant security issue in either of those ciphers is considered very unlikely, and future versions of Conscrypt/BoringSSL will continue to curate the list of supported cipher suites to make sure they're in good shape.)

Reply:
All good, this is definitely your call.  Just trying to understand, likely OkHttp will have to just log once that non-defaulted choices and order are being ignored.  

It does seem like TLS 1.3 still allows ordering and selection, just that you want that in conscrypt/GCM and updated more frequently than app developers configuring or OkHttp changing and then app developers upgrading.

Overall, this gives me most cause for concern.  Because it makes it impossible to workaround an implementation bug or security issue. But thanks for the headsup.

Reply:
Yep, the protocol still allows preference ordering, we're just choosing to determine the ordering ourselves instead of letting the caller customize it.  I agree that it's the most controversial change in our approach.

I previously e-mailed Jesse Wilson with a heads up on some of our choices, should I include you as well in the future?

Reply:
I would really appreciate it if you could cc the mailing list with this sort of information.

Reply:
Sure, we can do that in the future.

Reply:
I'm subscribed to conscrypt@googlegroups.com, so if it goes there, that's fine by me.  But I'm happy to be cc'd on any OkHttp+Conscrypt related side chats.